### ISSUE TITLE
JsonRpcApi input & output differs from C#

### ISSUE THREAD

**Issue by ixje** [CityOfZion/neo-python#233]
**issue 1**
C# accepts single quotes
```
{'params': ["ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9", "symbol"], 'id': '2', 'jsonrpc': '2.0', 'method': 'invokefunction'}
```
[JsonRpcApi](https://github.com/CityOfZion/neo-python/blob/development/examples/json-rpc-api-server.py) does not and will have an error saying:
```
{
    "jsonrpc": "2.0",
    "error": {
        "message": "Parse error",
        "code": -32700
    },
    "id": null
}
```
The actual error message is dropped and says
```
Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

**issue 2**
C# returns full names for the VMstate and Stack types after execution e.g.
```
http://seed3.cityofzion.io:8080  (2.7.0 node)

{'params': ["ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9", "symbol"], 'id': '2', 'jsonrpc': '2.0', 'method': 'invokefunction'}

{
    "jsonrpc": "2.0",
    "id": "2",
    "result": {
        "script": "00c10673796d626f6c67f91d6b7085db7c5aaf09f19eeec1ca3c0db2c6ec",
        "state": "HALT, BREAK",    <-----
        "gas_consumed": "0.141",
        "stack": [
            {
                "type": "ByteArray", <------
                "value": "525058"
            }
        ]
    }
}
```

JsonRpcApi gives VM state and stack type as (string) numbers to indicate the same. Note: the only difference in the calls is the scripthash I'm talking to.
```
{"id": "2", "method": "invokefunction", "params": ["584660c663f114d803928b77cb9abdd585a0fc17", "symbol"], "jsonrpc": "2.0"}

{
    "jsonrpc": "2.0",
    "result": {
        "script": "303063313036373337393664363236663663363731376663613038356435626439616362373738623932303364383134663136336336363034363538",
        "gas_consumed": "0.372",
        "stack": [
            {
                "value": "50434731",
                "type": "05" <--------
            }
        ],
        "state": 5 <-------
    },
    "id": "2"
}
```

**Comment by ixje** [comment]
I need to add one more difference in representation

C#
```
{'params': ["ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9", "decimals"], 'id': '2', 'jsonrpc': '2.0', 'method': 'invokefunction'}

            {
                "type": "Integer",
                "value": "8"  <---- string int
            }
```

`JsonRpcApi`

```
            {
                "value": 8,  <---- real in
                "type": "02"
            }
```

**Comment by metachris** [comment]
Very nice, detective @ixje.

**Comment by ixje** [comment]
**Issue 3**
C# accepts different input parameter types e.g. "Hash160" whereas `JsonRpcApi` would require "03" for the equivalent behaviour. ("Hash160") will throw exceptions

C#
```
{
	"jsonrpc": "2.0",
	"method": "invokefunction", 
	"params": [
    "ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9",
	"balanceOf",
	[
		{
        	"type": "Hash160", <-------
	        "value": "584660c663f114d803928b77cb9abdd585a0fc17"
    	}
	]
    ],
	"id": 2 
}
```

vs neo-python.
(note; again only script hashes swapped for privnet vs mainnet as i'm not synced up. Just focus on the arrows)

```
{
	"jsonrpc": "2.0",
	"method": "invokefunction", 
	"params": [
    "584660c663f114d803928b77cb9abdd585a0fc17",
	"balanceOf",
	[
		{
        	"type": "03", <------
	        "value": "ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9"
    	}
	]
    ],
	"id": 2 
}
```

**Comment by brianlenz** [comment]
**Issue 3** has been addressed in #267. I'm going to take a look at the other issues here in a bit.

**Comment by brianlenz** [comment]
For **issue 1**, I would actually argue that's a bug in the C# implementation. JSON doesn't support strings with single quotes, so I think getting an error in that case is justifiable.

Thoughts?

**Comment by brianlenz** [comment]
**issue 2** is fixed in #272, so the only outstanding question here is if we want to do anything for **issue 1** or not.

**Comment by metachris** [comment]
I think not, but I'm not religious about it. I think we can just stick with standard JSON. 

I did a quick look, and we'd probably have to switch to a custom "flexible" JSON decoder, rather than the standard ones.

**Comment by metachris** [comment]
Thanks so much @ixje and @brianlenz for documenting the efforts here. Closing this issue now, in favor of https://github.com/CityOfZion/neo-python/issues/273

**Comment by metachris** [comment]
Just a note for future readers, these issues are fixed on the current dev branch (`v0.4.11-dev`). Will report back once we merge it into master
